core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1251 @@
+    // into consideration.
+    let widthUpperBound = 548.6500000000001;
+
+    // The height of graph calculated as difference b/w bottom and right top.
+    let expectedHeight1 = expectedGraphBoundaries1.bottom -
+    expectedGraphBoundaries1.top;
+    let expectedHeight2 = expectedGraphBoundaries2.bottom -
+      expectedGraphBoundaries2.top;
+    let expectedHeight3 = expectedGraphBoundaries3.bottom -
+      expectedGraphBoundaries3.top;
+
+    // Here, we see that for 3 nodes the graph width is less than
+    // the upper bound, for  4 nodes the graph width is equal to the
+    // upper bound and the width becomes constant for nodes > 4.
+    expect(expectedWidth1).toBeLessThan(expectedWidth2);
+    expect(expectedWidth2).toEqual(widthUpperBound);
+    expect(expectedWidth3).toEqual(widthUpperBound);
+
+    // As height does not have an upper bound, it increases as a node overflows
+    // to the next row.
+
+    expect(expectedHeight1).toEqual(expectedHeight2);
+    expect(expectedHeight2).toBeLessThan(expectedHeight3);
+  });
+
+  it('should return graph boundaries with height equal to' +
+  ' the graph height', () => {
+    let nodeData = {
+      State1: {
+        depth: 1,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.42000000000000004,
+        xLabel: 0.1625,
+        yLabel: 0.5,
+        id: 'State1',
+        label: 'State1',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      Introduction: {
+        depth: 0,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.15333333333333335,
+        xLabel: 0.1625,
+        yLabel: 0.23333333333333334,
+        id: 'Introduction',
+        label: 'Introduction',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      End: {
+        depth: 2,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.6866666666666666,
+        xLabel: 0.1625,
+        yLabel: 0.7666666666666666,
+        id: 'End',
+        label: 'End',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State2: {
+        depth: 1,
+        offset: 1,
+        reachable: true,
+        x0: 0.29750000000000004,
+        y0: 0.42000000000000004,
+        xLabel: 0.3875,
+        yLabel: 0.5,
+        id: 'State2',
+        label: 'State2',
+        height: 0.16,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      }
+    };
+    let graphWidth = sgls.getGraphWidth(
+      AppConstants.MAX_NODES_PER_ROW, AppConstants.MAX_NODE_LABEL_LENGTH);
+    let graphHeight = sgls.getGraphHeight(nodeData);
+    let nodeDataWithPositionValueInPixel = sgls.modifyPositionValues(
+      nodeData, graphWidth, graphHeight);
+
+    let expectedGraphBoundaries = {
```

Comment:
There needs to be an explanation for why the expected values are what they are.

Reply:
Done